### ISSUE TITLE
Not displaying large numbers (>16 digits) correctly

### ISSUE THREAD

**Issue by titus-ong** [mwouts/itables#152]
The issue: Large numbers only display first 16 digits, 17th digit is rounded (sometimes incorrectly) and the rest are 0s.

How to replicate:
```py
from itables import show
import pandas as pd

a = pd.DataFrame(
    {
        "Name": [
            "Braund, Mr. Owen Harris",
            "Allen, Mr. William Henry",
            "Bonnell, Miss. Elizabeth",
        ],
        "Age": [1234567890123456789, 2345678901234567890, 3456789012345678901],
        "Sex": ["male", "male", "female"],
    }
)

show(a)
```

Result:

Name | Age | Sex
-- | -- | --
Braund, Mr. Owen Harris | 1234567890123456800 | male
Allen, Mr. William Henry | 2345678901234567700 | male
Bonnell, Miss. Elizabeth | 3456789012345679000 | female

Note that Allen's 17th digit (the second 7 in 77) should be a 9.

Version: 1.4.4

Notebook: Jupyter Notebook

**Comment by mwouts** [comment]
Oh interesting! Thanks for taking the time to report this. I will have a look and see how to fix this soon.

**Comment by titus-ong** [comment]
Cool, the issue of the 0's always appears for large numbers but I've not yet looked much into why the rounding sometimes doesn't work properly. For those facing this issue, you could either convert the column into string or have an additional column displaying the last few digits (e.g. using modulus).

**Comment by mwouts** [comment]
This is what I found:
- these long integers are supported by `json`:
```python
import json
json.dumps([1234567890123456789, 2345678901234567890, 3456789012345678901])
```
```
'[1234567890123456789, 2345678901234567890, 3456789012345678901]'
```
- however, these long integers are not valid in Javascript, so `datatables` [cannot display them properly](https://datatables.net/forums/discussion/55870/datatables-is-editing-my-long-integer)

I think we have two options: either convert to strings the numbers that are larger than `Number.MAX_SAFE_INTEGER`, or try using the `BigInt` javascript type. I will give a try to the second option following the comments [here](https://datatables.net/forums/discussion/67891/sorting-broken-with-big-numbers).

**Comment by mwouts** [comment]
I've done some research/experiments on `BigInt`, and unfortunately I have not been able to modify our [custom JSON encoder](https://github.com/mwouts/itables/blob/992a3f261e15a6f8c57aafa16daa3062d22ff5a6/itables/datatables_format.py#L30-L55) to automatically map the large Python integers to Javascript `BigInt`. In principle, we just need to add a `n` at the end of these large numbers, e.g. `[1234567890123456789n, 2345678901234567890n, 3456789012345678901n]` for the array described above. However in practice we cannot override the encoding of the known types, cf. [this SO question](https://stackoverflow.com/questions/46204503/why-is-my-custom-jsonencoder-default-ignoring-booleans).

I have not tested how well the `BigInt` work with `datatables`, so I am not sure whether this will work when we pass the step of the JSON encoding. What makes me doubt about this are the last comments about [sorting big numbers](https://datatables.net/forums/discussion/67891/sorting-broken-with-big-numbers). Maybe @AllanJard I can ask you whether you expect `BigInt` to work in datatables? If that helps I can ensure that these column only contains `BigInt` (no other integers types, but maybe a few strings like `"NA"`, `"Infinity"`).

**Comment by AllanJard** [comment]
The answer is no, the current release doesn't support BigInt. It throws an error in the number handling code. However, I've just [committed a couple of changes](https://github.com/DataTables/DataTablesSrc/commit/8138a66fde193a8cc27339093ee46e167592ba82) which addresses that issue, and it will now correctly render with a BigInt column.

**Comment by AllanJard** [comment]
Just to say - the other option is to make your long numbers strings in the JSON feed.

**Comment by mwouts** [comment]
> The answer is no, the current release doesn't support BigInt. It throws an error in the number handling code. However, I've just [committed a couple of changes](https://github.com/DataTables/DataTablesSrc/commit/8138a66fde193a8cc27339093ee46e167592ba82) which addresses that issue, and it will now correctly render with a BigInt column.

Oh that's interesting! In the long term I think this is the way to go

> Just to say - the other option is to make your long numbers strings in the JSON feed.

Yes I think that is the best short term approach. On the Python side I will convert the numbers that are larger than `MAX_SAFE_INTEGER` to strings. The only drawback I see is that sorting might not work well, but that I might display a (discardable) warning to make sure the user is aware of that.

**Comment by mwouts** [comment]
I see that [`datatables==1.13.2`](https://cdn.datatables.net/1.13.2/) added support for big ints (thank you @AllanJard !).

I still have some work to do on the Python side as well but I will give it a try.

**Comment by mwouts** [comment]
Well for the moment it seems more reasonable to convert the columns that contains big integers to string.

@titus-ong would you like to give a try to the proposed fix? You can install it with
```
pip install git+https://github.com/mwouts/itables.git@bigint_to_str
```

**Comment by titus-ong** [comment]
![image](https://user-images.githubusercontent.com/55629508/227762515-a782b50e-6836-45cc-8e42-dcb3327c391c.png)
Thanks for the fix, it works! Suppressing the warnings also work.

**Comment by mwouts** [comment]
Awesome! Thanks for confirming. By the way, I am wondering whether the default value for `warn_on_int_to_str_conversion` should be `True` or `False`. What do you think? As we don't lose much by converting to string, maybe it is not necessary to issue a warning for this? (only the sorting will not work properly after the conversion)

**Comment by titus-ong** [comment]
Ah I didn't think about sorting, but since it affects sorting I think the warning could be `True` by default. I don't require sorting for my use case but I can imagine that's one of your package's features that others may use a fair amount, so it might be better to warn pre-emptively. There probably wouldn't be a lot of users who deal with such large numbers so I don't think the warning is excessive either.

**Comment by mwouts** [comment]
Thanks for your feedback! I will let the warning by default then. This workaround will be available in `itables==1.5.2`. I have also created a follow-up issue to provide a proper fix for this, but I must say that it is a bit out of reach for now: #172. Thanks!
